cmd/admin-handlers-users.go

```diff
@@ -540,33 +545,29 @@ func (a adminAPIHandlers) AddServiceAccount(w http.ResponseWriter, r *http.Reque
 	}
 
 	var ldapUsername string
-	if globalLDAPConfig.Enabled && targetUser != "" {
-		// If LDAP enabled, service accounts need
-		// to be created only for LDAP users.
+	switch {
+	case owner && targetUser == cred.AccessKey:
+		// Allow service account creation for root user.
+	case globalLDAPConfig.Enabled:
+		// If LDAP enabled, service accounts need to be created only for
+		// LDAP users.
 		var err error
 		ldapUsername = targetUser
 		targetUser, targetGroups, err = globalLDAPConfig.LookupUserDN(targetUser)
 		if err != nil {
 			writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
 			return
 		}
-		// targerUser is set to bindDN at this point in time.
-		// targetGroups is set to the groups at this point in time.
-	} else {
+	case targetUser == cred.AccessKey:
 		if cred.IsServiceAccount() || cred.IsTemp() {
 			if cred.ParentUser == "" {
 				writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx,
 					errors.New("service accounts cannot be generated for temporary credentials without parent")), r.URL)
 				return
 			}
-			if targetUser == "" {
-				targetUser = cred.ParentUser
-			}
-		}
-		// targetGroups not yet set, so set this to cred.Groups
-		if len(targetGroups) == 0 {
-			targetGroups = cred.Groups
+			targetUser = cred.ParentUser
```

Comment:
Looks like this behavior is correct.

Reply:
Yes, it should preserve the earlier behaviour for all the previous cases. I tried to lay out the cases a bit more explicitly so we can confirm this code is working as expected.